src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -29,14 +29,15 @@ internal abstract class DictionaryDefaultConverter<TCollection, TKey, TValue>
         /// </summary>
         protected virtual void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state) { }
 
-        private static Type s_valueType = typeof(TValue);
+        private static readonly Type s_valueType = typeof(TValue);
 
         internal override Type ElementType => s_valueType;
 
-        protected Type KeyType = typeof(TKey);
+        protected static readonly Type KeyType = typeof(TKey);
+
         // For string keys we don't use a key converter
         // in order to avoid performance regression on already supported types.
-        protected bool IsStringKey = typeof(TKey) == typeof(string);
+        private static readonly bool s_isStringKey = KeyType == typeof(string);
```

Comment:
Generally speaking, when the `(x == typeof(...))` branch is false, that whole code path is optimized out. Will  that remain true with this pattern as well, along with how we use `ElementType`?

For example:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L434-L455

[sharplab.io example](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4CS51ATgPYAOAyswG4UDG1CAG50WXARLkqtAMIsobCgBtmXJrwHDRODgAswTNgBkwAIyJlKNEWlEBmcQDZcCbDOwBvdNh/ZvvgG0AKQpgAHFqUmZ+AApgAE82ahYAMxjTeOBqAEpsgBpsEPDI6L44xOS0vn0mXIKiiKimWISk1Ji8TFyAXX8fOAc8ZwpybC4AR2JIgTkFA2oAFRYAHgWAPjjdCggxtjBSVbXsCD3SAsJqMAATAHlSJXiOU8PsFmBdZmy+3x+fAHcPkxqNgFtgQNg6LM9kwzCpDt8vGhfr4KClsOU2mkFtlsABeXHYVqVdKZHJfJHI35wADsY2ok2m1Ch8yWMW+lN+QLRAFVSBAwClaABBCCrAoZLIbLnYACy1CgLCY8RlBgg+iU+AiwAugqBpAEMRO+zq7I5viNpHwRkiAHN3nlTWbsNLefzBfgRWLsBLqFLqGi5QqlSqmGqwBqtTrmIyYm9ASaKU6fnHmFbbe9sjZHdhUeiie1sXiCfmqjVctmfjS6Qz9Uz5NDFiw2Ymkz4XXyBcLRQsCtUDH6A/LFcrVerNdRtf7o7XDacE62fha06Q7boHS3W+23V2vX2mAPZUPg6Pw+PJ7qYynavkK5Sr8vV5n0NmqxMprXmUDWdLA8OQ2GIwnKM9QNC18mOU4H3tZ1/UPIMR1DMdIynEDqFjd5PgKe9rRXDMbB+ABfZ8WwCOV3hYK46AUJQYjI3QKKotglBuNhKBYPl8AAORYBglBGEYbWyXoWzYZpuDALInBzUY30ZT9GxiaUfW9JAChGYBsCUVSYLRZTTAQNTRiUBByR+REOSrTB8N8IiW2+UiJ3oyjqNoxyGOoli2I47jeP4lchO+USKHEyShmkjTZI/esWSbaU9xUwyNK0go4pqb0DPCzSTIRF9aSs75bJ+ey6PcpjXPI5ymM8ih2IgLiePuPzBOEn4gpC4EwvU6t3xmaKvxWdZFNg0FTG0rrkp0kF0sSrLTN8czKUs6yfFsgigA==)

Reply:
Local benchmarks show using `(x == typeof(...))` can be faster and has potentially lower code size than using a cached `Type` instance. I made a minor change in this PR to align better with this, but plan to do an audit of how we use `typeof()` in a follow up PR backed with microbenchmarks/JSON benchmarks.